File: src/engine/layout.ts
/**
 * Arranging a machine so the diagram reads.
 *
 * The old Tidy put every state on a ring, and a ring is the wrong shape twice
 * over.
 *
 * It is wrong in the small. The radius came from a circumference estimate,
 * `n * spacing / 2π`, which treats the arc between two neighbours as if it
 * were the straight line between them. For three states the arc is 21% longer
 * than the chord, so three states of radius 28 landed on a ring of radius 36,
 * six units of clear air between their rims. The circles themselves just about
 * miss each other; a self loop does not. One reaches 1.82 radii out from its
 * own centre, so on that ring every loop was drawn straight through the state
 * next to it, and every three state level has loops.
 *
 * It is wrong in the large too: almost every machine here is a chain, and a
 * chain drawn on a circle reads as a circle, not as a sequence.
 *
 * So rank the states by how far they are from the start along the arrows, and
 * give each rank a column. A chain comes out as a row, in the order it runs. A
 * branch comes out as a fork. Within a column, order by the average height of
 * whatever points at a state, which is the cheap half of what a real graph
 * layout engine does about edge crossings, and at four or five states it is the
 * half that matters.
 *
 * Nothing here is clamped to the logical canvas. Tidy is always followed by a
 * fit, which frames whatever it produced, and a six state chain simply needs
 * more room than a 340 unit wide coordinate frame has.
 */

import { CANVAS, type Machine, type StateId } from './types';

export interface Point {
  x: number;
  y: number;
}

const R = CANVAS.stateRadius;

/** Room between columns for an arrow and the chip riding on it. */
const COL_PITCH = 2 * R + 66;

/** Room between rows for a self loop clear of the state above or below it. */
const ROW_PITCH = 2 * R + 48;

/** Above this, a rank with no structure to it is better off as a grid. */
const GRID_ABOVE = 3;

/** Lay out `n` states with nothing joining them: a squarish grid, centred. */
function gridLayout(n: number): Point[] {
  const cols = Math.ceil(Math.sqrt(n));
  const rows = Math.ceil(n / cols);
  const x0 = CANVAS.width / 2 - ((cols - 1) * COL_PITCH) / 2;
  const y0 = CANVAS.height / 2 - ((rows - 1) * ROW_PITCH) / 2;
  return Array.from({ length: n }, (_, i) => ({
    x: Math.round(x0 + (i % cols) * COL_PITCH),
    y: Math.round(y0 + Math.floor(i / cols) * ROW_PITCH),
  }));
}

/**
 * Positions for every state of `m`, in the order `m.states` is in.
 *
 * Reads only the graph, never the current coordinates, so it is idempotent:
 * tidying twice gives the same picture as tidying once.
 */
export function layoutMachine(m: Machine): Point[] {
  const n = m.states.length;
  if (n === 0) return [];
  if (n === 1) return [{ x: CANVAS.width / 2, y: CANVAS.height / 2 }];

  const index = new Map<StateId, number>(m.states.map((s, i) => [s.id, i]));

  const out: number[][] = m.states.map(() => []);
  const into: number[][] = m.states.map(() => []);
  for (const t of m.transitions) {
    const a = index.get(t.from);
    const b = index.get(t.to);
    // A self loop says nothing about where a state belongs.
    if (a === undefined || b === undefined || a === b) continue;
    out[a]?.push(b);
    into[b]?.push(a);
  }

  // Rank by breadth first search along the arrows. The start state goes first
  // so the machine is laid out in the direction it runs; anything the start
  // cannot reach is rooted separately, at rank 0, so a stranded state sits
  // beside the beginning rather than trailing off the end.
  const rank = new Array<number>(n).fill(-1);
  const startIndex = m.start === null ? undefined : index.get(m.start);
  const roots = [...(startIndex === undefined ? [] : [startIndex])];
  for (let i = 0; i < n; i++) if (i !== startIndex) roots.push(i);

  for (const root of roots) {
    if (rank[root] !== -1) continue;
    rank[root] = 0;
    const queue = [root];
    for (let head = 0; head < queue.length; head++) {
      const u = queue[head] as number;
      for (const v of out[u] ?? []) {
        if (rank[v] !== -1) continue;
        rank[v] = (rank[u] as number) + 1;
        queue.push(v);
      }
    }
  }

  const depth = Math.max(...rank) + 1;
  // One rank and more than a handful means there is no sequence to show: a
  // scatter of unconnected states, or a set that all loop on themselves.
  if (depth === 1 && n > GRID_ABOVE) return gridLayout(n);

  const columns: number[][] = Array.from({ length: depth }, () => []);
  for (let i = 0; i < n; i++) (columns[rank[i] as number] as number[]).push(i);

  // Order within each column by the mean height of its predecessors in the
  // column before. Three sweeps is enough at this size and cannot oscillate
  // for long; the original order breaks ties, and that is usually the order
  // the states were drawn in.
  const row = new Array<number>(n).fill(0);
  for (const col of columns) col.forEach((i, j) => (row[i] = j));

  for (let pass = 0; pass < 3; pass++) {
    for (let c = 1; c < columns.length; c++) {
      const col = columns[c] as number[];
      const key = new Map<number, number>();
      for (const i of col) {
        const above = (into[i] ?? []).filter((p) => rank[p] === c - 1);
        key.set(
          i,
          above.length === 0
            ? (row[i] as number)
            : above.reduce((sum, p) => sum + (row[p] as number), 0) / above.length,
        );
      }
      col.sort((a, b) => (key.get(a) as number) - (key.get(b) as number) || (row[a] as number) - (row[b] as number));
      col.forEach((i, j) => (row[i] = j));
    }
  }

  const x0 = CANVAS.width / 2 - ((depth - 1) * COL_PITCH) / 2;
  const midY = CANVAS.height / 2;
  const points = new Array<Point>(n);

  columns.forEach((col, c) => {
    // Every column is centred on the same line, so a chain is level and a
    // fork opens symmetrically about it.
    const top = midY - ((col.length - 1) * ROW_PITCH) / 2;
    col.forEach((i, j) => {
      points[i] = {
        x: Math.round(x0 + c * COL_PITCH),
        y: Math.round(top + j * ROW_PITCH),
      };
    });
  });

  return points;
}

/** `m` with its states moved to where `layoutMachine` puts them. */
export function laidOut(m: Machine): Machine {
  const points = layoutMachine(m);
  return {
    ...m,
    states: m.states.map((s, i) => ({ ...s, ...(points[i] ?? { x: s.x, y: s.y }) })),
  };
}

File: ios/Sources/AutomataEngine/Layout.swift
import Foundation

/// Arranging a machine so the diagram reads. A port of src/engine/layout.ts,
/// kept identical to it so a machine tidied on the phone and the same machine
/// tidied in the browser come out in the same shape.
///
/// Rank the states by how far they are from the start along the arrows and
/// give each rank a column: a chain comes out as a row in the order it runs, a
/// branch comes out as a fork. Within a column, order by the average height of
/// whatever points at a state — the cheap half of what a real graph layout
/// engine does about edge crossings, and at five states the half that matters.
///
/// Nothing here is clamped to the logical canvas. Tidy is always followed by a
/// fit, and a six state chain needs more room than 340 units.

public struct Point: Sendable, Hashable {
    public var x: Double
    public var y: Double

    public init(x: Double, y: Double) {
        self.x = x
        self.y = y
    }
}

public enum Layout {

    private static let R = Canvas.stateRadius

    /// Least room between columns: an arrow, and a one character chip on it.
    private static let colPitch = 2 * R + 66

    /// Least room between rows: a self loop clear of the state above or below.
    private static let rowPitch = 2 * R + 48

    /// Advance of one character in the chip's face, and the height of one chip
    /// in a stack. Duplicated from the drawing code rather than imported: the
    /// engine does not depend on the UI.
    private static let chipCharW = 7.2
    private static let chipRowH = 24.0

    /// Above this, a rank with no structure to it is better off as a grid.
    private static let gridAbove = 3

    /// The class a machine's rules imply. A layout is never told the level.
    public static func kindOf(_ m: Machine) -> MachineKind {
        for t in m.transitions {
            if t.move != nil || t.write != nil { return .TM }
            if t.pop != nil || t.push != nil { return .PDA }
        }
        return .DFA
    }

    /// What a rule writes on its arrow. The app draws exactly this string.
    public static func chip(_ t: Transition, kind: MachineKind) -> String {
        switch kind {
        case .DFA, .NFA:
            return t.read
        case .PDA:
            return "\(t.read), \(t.pop ?? Symbols.epsilon) → \(t.push ?? Symbols.epsilon)"
        case .TM:
            return "\(t.read) → \(t.write ?? t.read), \(t.move?.rawValue ?? "R")"
        }
    }

    /// Pitches wide enough for what this machine actually writes on its arrows.
    /// Spacing every machine as though it were a DFA is what left the dense
    /// levels with their chips lying on top of each other after a tidy.
    private static func pitches(_ m: Machine) -> (col: Double, row: Double) {
        let kind = kindOf(m)
        var widest = 1
        var perEdge: [String: Int] = [:]
        for t in m.transitions {
            widest = max(widest, chip(t, kind: kind).count)
            let key = "\(t.from)\u{0}\(t.to)"
            perEdge[key, default: 0] += 1
        }
        let stacked = max(1, perEdge.values.max() ?? 1)
        return (
            col: max(colPitch, 2 * R + Double(widest) * chipCharW + 30),
            row: max(rowPitch, 2 * R + Double(stacked) * chipRowH + 24)
        )
    }

    /// Lay out `n` states with nothing joining them: a squarish grid, centred.
    private static func gridLayout(_ n: Int, col: Double, row: Double) -> [Point] {
        let cols = max(1, Int(ceil(Double(n).squareRoot())))
        let rows = Int(ceil(Double(n) / Double(cols)))
        let x0 = Canvas.width / 2 - (Double(cols - 1) * col) / 2
        let y0 = Canvas.height / 2 - (Double(rows - 1) * row) / 2
        return (0..<n).map { i in
            Point(
                x: (x0 + Double(i % cols) * col).rounded(),
                y: (y0 + Double(i / cols) * row).rounded()
            )
        }
    }

    /// Positions for every state of `m`, in the order `m.states` is in.
    ///
    /// Reads only the graph, never the current coordinates, so it is
    /// idempotent: tidying twice gives the same picture as tidying once.
    public static func layoutMachine(_ m: Machine) -> [Point] {
        let n = m.states.count
        if n == 0 { return [] }
        if n == 1 { return [Point(x: Canvas.width / 2, y: Canvas.height / 2)] }

        let (COL, ROW) = pitches(m)
        var index: [String: Int] = [:]
        for (i, s) in m.states.enumerated() { index[s.id] = i }

        var out: [[Int]] = Array(repeating: [], count: n)
        var into: [[Int]] = Array(repeating: [], count: n)
        for t in m.transitions {
            guard let a = index[t.from], let b = index[t.to], a != b else { continue }
            out[a].append(b)
            into[b].append(a)
        }

        // Rank by breadth first search along the arrows. The start goes first
        // so the machine is laid out in the direction it runs; anything the
        // start cannot reach is rooted separately at rank 0, so a stranded
        // state sits beside the beginning rather than trailing off the end.
        var rank = Array(repeating: -1, count: n)
        let startIndex = m.start.flatMap { index[$0] }
        var roots: [Int] = []
        if let s = startIndex { roots.append(s) }
        for i in 0..<n where i != startIndex { roots.append(i) }

        for root in roots {
            if rank[root] != -1 { continue }
            rank[root] = 0
            var queue = [root]
            var head = 0
            while head < queue.count {
                let u = queue[head]
                head += 1
                for v in out[u] where rank[v] == -1 {
                    rank[v] = rank[u] + 1
                    queue.append(v)
                }
            }
        }

        let depth = (rank.max() ?? 0) + 1
        // One rank and more than a handful means there is no sequence to show.
        if depth == 1 && n > gridAbove { return gridLayout(n, col: COL, row: ROW) }

        var columns: [[Int]] = Array(repeating: [], count: depth)
        for i in 0..<n { columns[rank[i]].append(i) }

        // Order within each column by the mean height of its predecessors in
        // the column before. Three sweeps is enough at this size; the original
        // order breaks ties, and that is usually the order they were drawn in.
        var row = Array(repeating: 0, count: n)
        for col in columns {
            for (j, i) in col.enumerated() { row[i] = j }
        }

        for _ in 0..<3 {
            for c in 1..<columns.count {
                var key: [Int: Double] = [:]
                for i in columns[c] {
                    let above = into[i].filter { rank[$0] == c - 1 }
                    key[i] = above.isEmpty
                        ? Double(row[i])
                        : above.reduce(0.0) { $0 + Double(row[$1]) } / Double(above.count)
                }
                columns[c].sort { a, b in
                    let ka = key[a] ?? 0
                    let kb = key[b] ?? 0
                    if ka != kb { return ka < kb }
                    return row[a] < row[b]
                }
                for (j, i) in columns[c].enumerated() { row[i] = j }
            }
        }

        let x0 = Canvas.width / 2 - (Double(depth - 1) * COL) / 2
        let midY = Canvas.height / 2
        var points = Array(repeating: Point(x: 0, y: 0), count: n)

        for (c, column) in columns.enumerated() {
            // Every column is centred on the same line, so a chain is level
            // and a fork opens symmetrically about it.
            let top = midY - (Double(column.count - 1) * ROW) / 2
            for (j, i) in column.enumerated() {
                points[i] = Point(
                    x: (x0 + Double(c) * COL).rounded(),
                    y: (top + Double(j) * ROW).rounded()
                )
            }
        }

        return points
    }

    /// `m` with its states moved to where `layoutMachine` puts them.
    public static func laidOut(_ m: Machine) -> Machine {
        let points = layoutMachine(m)
        var next = m
        for (i, p) in points.enumerated() where i < next.states.count {
            next.states[i].x = p.x
            next.states[i].y = p.y
        }
        return next
    }
}
